src/Effectful/HPQTypes.hs

```diff
@@ -71,72 +79,128 @@ runEffectDB ::
+    GetLastQuery -> unDBEff PQ.getLastQuery
+    GetTransactionSettings -> unDBEff PQ.getTransactionSettings
+    SetTransactionSettings settings -> unDBEff $ PQ.setTransactionSettings settings
+    WithFrozenLastQuery (action :: Eff localEs b) ->
+      unDBEff . PQ.withFrozenLastQuery . DBEff $
+        localSeqUnlift env $
+          \unlift -> unlift action
+    WithNewConnection (action :: Eff localEs b) ->
+      unDBEff . PQ.withNewConnection . DBEff $
+        localSeqUnlift env $
+          \unlift -> unlift action
+    GetNotification time -> unDBEff $ PQ.getNotification time
   where
-    runWithState :: Eff (State (PQ.DBState (Eff es)) : es) a -> Eff es a
+    runWithState :: Eff (DBState es : es) a -> Eff es a
     runWithState eff =
       PQ.withConnection (PQ.unConnectionSource connectionSource) $ \conn -> do
-        let dbState0 = mkDBState conn transactionSettings
-        evalState dbState0 eff :: Eff es a
-    runWithNewConnection ::
-      Eff (State (PQ.DBState (Eff es)) : es) b ->
-      Eff (State (PQ.DBState (Eff es)) : es) b
-    runWithNewConnection action = do
-      dbState :: PQ.DBState (Eff es) <- get
-      result <- PQ.withConnection (PQ.unConnectionSource connectionSource) $ \newConn -> do
-        -- TODO: We do not pass the current connection source to the new DB
-        -- state, which differs from the original code.
-        put $ mkDBState newConn (PQ.dbTransactionSettings dbState)
-        action
-      put dbState
-      pure result
-    mkDBState :: PQ.Connection -> PQ.TransactionSettings -> PQ.DBState (Eff es)
-    mkDBState conn ts =
-      PQ.DBState
-        { PQ.dbConnection = conn
-        , PQ.dbConnectionSource = PQ.unConnectionSource connectionSource
-        , PQ.dbTransactionSettings = ts
-        , PQ.dbLastQuery = PQ.SomeSQL (mempty :: PQ.SQL)
-        , PQ.dbRecordLastQuery = True
-        , PQ.dbQueryResult = Nothing
-        }
+        let dbState0 = mkDBState (PQ.unConnectionSource connectionSource) conn transactionSettings
+        evalState dbState0 $ handleAutoTransaction transactionSettings withTransaction' eff
+    withTransaction' ::
+      PQ.TransactionSettings ->
+      Eff (DBState es : es) a ->
+      Eff (DBState es : es) a
+    withTransaction' ts eff = unDBEff . PQ.withTransaction' ts $ DBEff eff
+
+mkDBState ::
+  PQ.ConnectionSourceM m ->
+  PQ.Connection ->
+  PQ.TransactionSettings ->
+  PQ.DBState m
+mkDBState connectionSource conn ts =
+  PQ.DBState
+    { PQ.dbConnection = conn
+    , PQ.dbConnectionSource = connectionSource
+    , PQ.dbTransactionSettings = ts
+    , PQ.dbLastQuery = PQ.SomeSQL (mempty :: PQ.SQL)
+    , PQ.dbRecordLastQuery = True
+    , PQ.dbQueryResult = Nothing
+    }
+
+handleAutoTransaction ::
+  PQ.TransactionSettings ->
+  (PQ.TransactionSettings -> m a -> m a) ->
+  m a ->
+  m a
+handleAutoTransaction transactionSettings withTransaction action =
+  -- REVIEW: Why don't we have to set `tsAutoTransaction` to `False` in the
```

Comment:
Because if the action calls `commit` inside, then with `tsAutoTransaction` another transaction should be started automatically and if it's not set, it won't happen (see source of the `commit'` function).

On the other hand, `withTransaction` itself uses these functions and there you don't want to start another transaction. I can see that the way it's implemented is somewhat confusing though.

Reply:
Thanks for the explanation!  I added it as a comment in the code.